File: k2/csrc/arcsort_test.cc
// k2/csrc/arcsort_test.cc

// Copyright (c)  2020  Xiaomi Corporation (author: Haowen Qiu)

// See ../../LICENSE for clarification regarding multiple authors

#include "k2/csrc/arcsort.h"

#include <algorithm>
#include <memory>
#include <utility>
#include <vector>

#include "gmock/gmock.h"
#include "gtest/gtest.h"
#include "k2/csrc/fsa.h"
#include "k2/csrc/fsa_util.h"
#include "k2/csrc/properties.h"

namespace k2 {
TEST(ArcSortTest, ArcSorter) {
  // empty fsa
  {
    // case 1: empty input fsa
    FsaCreator fsa_creator;
    const auto &fsa = fsa_creator.GetFsa();
    ArcSorter sorter(fsa);
    Array2Size<int32_t> fsa_size;
    sorter.GetSizes(&fsa_size);
    EXPECT_EQ(fsa_size.size1, 0);
    EXPECT_EQ(fsa_size.size2, 0);

    FsaCreator fsa_creator_out(fsa_size);
    auto &arc_sorted = fsa_creator_out.GetFsa();
    std::vector<int32_t> arc_map(fsa_size.size2);
    sorter.GetOutput(&arc_sorted, arc_map.data());

    EXPECT_TRUE(IsEmpty(arc_sorted));
    EXPECT_TRUE(arc_map.empty());
  }

  {
    std::vector<Arc> src_arcs = {
        {0, 1, 2}, {0, 4, 0}, {0, 2, 0}, {1, 2, 1}, {1, 3, 0}, {2, 1, 0},
    };
    FsaCreator fsa_creator(src_arcs, 4);
    const auto &fsa = fsa_creator.GetFsa();
    ArcSorter sorter(fsa);
    Array2Size<int32_t> fsa_size;
    sorter.GetSizes(&fsa_size);

    FsaCreator fsa_creator_out(fsa_size);
    auto &arc_sorted = fsa_creator_out.GetFsa();
    std::vector<int32_t> arc_map(fsa_size.size2);
    sorter.GetOutput(&arc_sorted, arc_map.data());

    EXPECT_FALSE(arc_map.empty());
    EXPECT_TRUE(IsArcSorted(arc_sorted));

    std::vector<int32_t> arc_indexes(arc_sorted.indexes,
                                     arc_sorted.indexes + arc_sorted.size1 + 1);
    std::vector<Arc> arcs(arc_sorted.data, arc_sorted.data + arc_sorted.size2);
    EXPECT_THAT(arc_indexes, ::testing::ElementsAre(0, 3, 5, 6, 6, 6));
    ASSERT_EQ(arcs.size(), fsa.size2);
    std::vector<Arc> target_arcs = {
        {0, 2, 0}, {0, 4, 0}, {0, 1, 2}, {1, 3, 0}, {1, 2, 1}, {2, 1, 0},
    };
    for (std::size_t i = 0; i != target_arcs.size(); ++i)
      EXPECT_EQ(arcs[i], target_arcs[i]);

    // arc index in `arc_sortd` -> arc index in original `fsa`
    // 0 -> 2
    // 1 -> 1
    // 2 -> 0
    // 3 -> 4
    // 4 -> 3
    // 5 -> 5
    EXPECT_THAT(arc_map, ::testing::ElementsAre(2, 1, 0, 4, 3, 5));
  }
}

TEST(ArcSortTest, ArcSort) {
  // empty fsa
  {
    // case 1: empty input fsa
    FsaCreator fsa_creator;
    auto &fsa = fsa_creator.GetFsa();
    std::vector<int32_t> arc_map(fsa.size2);
    ArcSort(&fsa, arc_map.data());

    EXPECT_TRUE(IsEmpty(fsa));
    EXPECT_TRUE(arc_map.empty());
  }

  {
    std::vector<Arc> src_arcs = {
        {0, 1, 2}, {0, 4, 0}, {0, 2, 0}, {1, 2, 1}, {1, 3, 0}, {2, 1, 0},
    };
    FsaCreator fsa_creator(src_arcs, 4);
    auto &fsa = fsa_creator.GetFsa();
    std::vector<int32_t> arc_map(fsa.size2);
    ArcSort(&fsa, arc_map.data());

    EXPECT_TRUE(IsArcSorted(fsa));

    std::vector<int32_t> arc_indexes(fsa.indexes, fsa.indexes + fsa.size1 + 1);
    std::vector<Arc> arcs(fsa.data, fsa.data + fsa.size2);
    EXPECT_THAT(arc_indexes, ::testing::ElementsAre(0, 3, 5, 6, 6, 6));
    std::vector<Arc> target_arcs = {
        {0, 2, 0}, {0, 4, 0}, {0, 1, 2}, {1, 3, 0}, {1, 2, 1}, {2, 1, 0},
    };
    for (std::size_t i = 0; i != target_arcs.size(); ++i)
      EXPECT_EQ(arcs[i], target_arcs[i]);

    // arc index in `arc_sortd` -> arc index in original `fsa`
    // 0 -> 2
    // 1 -> 1
    // 2 -> 0
    // 3 -> 4
    // 4 -> 3
    // 5 -> 5
    EXPECT_THAT(arc_map, ::testing::ElementsAre(2, 1, 0, 4, 3, 5));
  }
}
}  // namespace k2

File: k2/python/host/tests/arcsort_test.py
#!/usr/bin/env python3
#
# Copyright (c)  2020  Xiaomi Corporation (author: Haowen Qiu)
#
# See ../../../LICENSE for clarification regarding multiple authors

# To run this single test, use
#
#  ctest --verbose -R host_arcsort_test_py
#

import unittest

import torch

import k2host
from k2host.fsa_util import float_to_int


class TestArcSort(unittest.TestCase):

    def test_empty_fsa(self):
        array_size = k2host.IntArray2Size(0, 0)
        fsa = k2host.Fsa.create_fsa_with_size(array_size)
        arc_map = k2host.IntArray1.create_array_with_size(fsa.size2)
        k2host.arc_sort(fsa, arc_map)
        self.assertTrue(k2host.is_empty(fsa))
        self.assertTrue(arc_map.empty())

        # test without arc_map
        k2host.arc_sort(fsa)
        self.assertTrue(k2host.is_empty(fsa))

    def test_arc_sort(self):
        s = r'''
        0 1 2 1
        0 4 0 2
        0 2 0 3
        1 2 1 4
        1 3 0 5
        2 1 0 6
        4
        '''

        fsa = k2host.str_to_fsa(s)
        arc_map = k2host.IntArray1.create_array_with_size(fsa.size2)
        k2host.arc_sort(fsa, arc_map)
        expected_arc_indexes = torch.IntTensor([0, 3, 5, 6, 6, 6])
        expected_arcs = torch.IntTensor([[0, 2, 0, float_to_int(3)],
                                         [0, 4, 0, float_to_int(2)],
                                         [0, 1, 2, float_to_int(1)],
                                         [1, 3, 0, float_to_int(5)],
                                         [1, 2, 1, float_to_int(4)],
                                         [2, 1, 0, float_to_int(6)]])
        expected_arc_map = torch.IntTensor([2, 1, 0, 4, 3, 5])
        self.assertTrue(torch.equal(fsa.indexes, expected_arc_indexes))
        self.assertTrue(torch.equal(fsa.data, expected_arcs))
        self.assertTrue(torch.equal(arc_map.data, expected_arc_map))


class TestArcSorter(unittest.TestCase):

    def test_empty_fsa(self):
        array_size = k2host.IntArray2Size(0, 0)
        fsa = k2host.Fsa.create_fsa_with_size(array_size)
        sorter = k2host.ArcSorter(fsa)
        array_size = k2host.IntArray2Size()
        sorter.get_sizes(array_size)
        fsa_out = k2host.Fsa.create_fsa_with_size(array_size)
        arc_map = k2host.IntArray1.create_array_with_size(array_size.size2)
        sorter.get_output(fsa_out, arc_map)
        self.assertTrue(k2host.is_empty(fsa))

        # test without arc_map
        sorter.get_output(fsa_out)
        self.assertTrue(k2host.is_empty(fsa_out))

    def test_arc_sort(self):
        s = r'''
        0 1 2 1
        0 4 0 2
        0 2 0 3
        1 2 1 4
        1 3 0 5
        2 1 0 6
        4
        '''

        fsa = k2host.str_to_fsa(s)
        sorter = k2host.ArcSorter(fsa)
        array_size = k2host.IntArray2Size()
        sorter.get_sizes(array_size)
        fsa_out = k2host.Fsa.create_fsa_with_size(array_size)
        arc_map = k2host.IntArray1.create_array_with_size(array_size.size2)
        sorter.get_output(fsa_out, arc_map)
        expected_arc_indexes = torch.IntTensor([0, 3, 5, 6, 6, 6])
        expected_arcs = torch.IntTensor([[0, 2, 0, float_to_int(3)],
                                         [0, 4, 0, float_to_int(2)],
                                         [0, 1, 2, float_to_int(1)],
                                         [1, 3, 0, float_to_int(5)],
                                         [1, 2, 1, float_to_int(4)],
                                         [2, 1, 0, float_to_int(6)]])
        expected_arc_map = torch.IntTensor([2, 1, 0, 4, 3, 5])
        self.assertTrue(torch.equal(fsa_out.indexes, expected_arc_indexes))
        self.assertTrue(torch.equal(fsa_out.data, expected_arcs))
        self.assertTrue(torch.equal(arc_map.data, expected_arc_map))


if __name__ == '__main__':
    unittest.main()
